### ISSUE TITLE
DateTime.Kind comes back as Unspecified

### ISSUE THREAD

**Issue by gzak** [dotnet/efcore#4711]
If you store a `DateTime` object to the DB with a `DateTimeKind` of either `Utc` or `Local`, when you read that record back from the DB you'll get a `DateTime` object whose kind is `Unspecified`. Basically, it appears the `Kind` property isn't persisted. This means that a test to compare equality of the before/after value of a round trip to the DB will fail.

Is there something that needs to be configured for the `Kind` to be preserved? Or what's the recommended way to handle this?

**Comment by gzak** [comment]
```
var entity = db.Entity.Single(e => e.Id == 123);
var now = DateTime.UtcNow;
entity.Date = now;
db.SaveChanges();

entity = db.Entity.Single(e => e.Id == 123);
Assert.Equal(now, entity.Date); // fails the assert
```

**Comment by rmja** [comment]
A slight variation of this is that it would be nice if one could configure the default `DateTimeKind`. For example, I know that all dates in my database are Utc, so it would be nice if this was specified for all fetched dates in the DbContext, or maybe configurable per entity property?

**Comment by ilmax** [comment]
+1 on whar @rmja suggested, it's a really nice addition.

**Comment by gzak** [comment]
One thing worth noting is that I think the `DateTime` type in C♯ has more precision than the DB equivalent, so equality probably won't be preserved after a round trip even with this change.

**Comment by ajcvickers** [comment]
Note for triage: Investigated mapping to DateTime CLR type to SQL Server datetimeoffset by default, but even when switching to the typed value buffer factory this still won't work without type conversions.

**Comment by rowanmiller** [comment]
Conclusions from triage:
* This is fundamentally a limitation of SQL Server/SQL Client, where the kind of datetime is not persisted.
* We could work around this by introducing a separate property to persist kind... but there is already `DateTimeOffset` which is designed to specifically store date/time in a time zone aware way. So really the solution isn't to try and workaround the limitations of `DateTime`, but educate folks to use `DateTimeOffset`.

**Comment by rmja** [comment]
@rowanmiller how about my suggestion? It would be really useful in systems where the kind does not need to be stored, because it is agreed upon by convention.

**Comment by rowanmiller** [comment]
@rmja I should have mentioned that too... at the moment we rely on ADO.NET to construct the value and then we set it on the property. We do want to introduce a feature where you can control how that value is get/set from/to the property. This would allow things like setting DateTime.Kind, type conversions, using Get/Set methods rather than properties, etc. https://github.com/aspnet/EntityFramework/issues/240 is the issue tracking this feature

**Comment by gzak** [comment]
@rowanmiller would you mind giving an example (like my second comment) that uses `DateTimeOffset` to correct the issue?

**Comment by gzak** [comment]
Never mind my last comment, I didn't realize there was a corresponding sql type called `datetimeoffset`. Basically, we've switched everything over to the `-offset` variants in both C# and the DB which solved this issue as well as #5175.

**Comment by rowanmiller** [comment]
@gzak glad it's working for you now

**Comment by hikalkan** [comment]
I also would want such a convention @rmja mentioned (https://github.com/aspnet/EntityFramework/issues/4711#issuecomment-201405020). Because not all database providers have datetimeoffset type.

**Comment by hikalkan** [comment]
I found a solution for that and shared as a blog post: http://volosoft.com/datetime-specifykind-in-entity-framework-core-while-querying/

**Comment by haavardheggheim** [comment]
I had this same issue today. Why is Datetime/Timezones still an issue? I don't really care how DateTime is stored in the database as long as the returned value is the same as when I stored it. It could even be stored as a standardized(ISO 8601) string for all I care.

**Comment by EEaglehouse** [comment]
Switching to DateTimeOffset doesn't solve the problem. It isn't time zone aware; it only stores the UTC offset for a specific instant in time. There is no simple way to derive the time zone where the value was created. Picking the underlying UTC value out of the DateTimeOffset value would be the best you could do and DateTime is just as good at assuming that.

**Comment by ajcvickers** [comment]
@EEaglehouse and others. I'm working on a change that will allow the kind to be be preserved by storing the DateTime as a long in the database and automatically converting it on the way in and out. Should be in 2.1 as part of #242.

**Comment by EEaglehouse** [comment]
@ajcvickers Thank you. It would be nice to have that as an option. For me, the better solution remains to have an option to force the Kind to a particular value (Utc in my case) for DateTime values. Reading it as a DateTimeOffset is useless to me because I would still need to convert to DateTime with Kind as Utc in my applications' data layer; all local time zone conversions are handled in the presentation layer. Converting DateTime to a long for storage may be useful, but would effectively obfuscate the column type in the database. But please don't take this as discouraging you from implementing the custom type mapping, because then I could do for myself what I'm asking.

**Comment by ajcvickers** [comment]
@EEaglehouse Yes, after #242 you will be able to set things up to still store as a datetime in the database, but force a specific kind whenever it is read from the database.

**Comment by werwolfby** [comment]
We are using this updated workaround from: http://volosoft.com/datetime-specifykind-in-entity-framework-core-while-querying/

```c#
    public class DateTimeKindMapper
    {
        public static DateTime Normalize(DateTime value)
            => DateTime.SpecifyKind(value, DateTimeKind.Utc);

        public static DateTime? NormalizeNullable(DateTime? value)
            => value.HasValue ? DateTime.SpecifyKind(value.Value, DateTimeKind.Utc) : (DateTime?)null;

        public static object NormalizeObject(object value)
            => value is DateTime dateTime ? Normalize(dateTime) : value;
    }

    public class DateTimeKindEntityMaterializerSource : EntityMaterializerSource
    {
        private static readonly MethodInfo _normalizeMethod =
            typeof(DateTimeKindMapper).GetTypeInfo().GetMethod(nameof(DateTimeKindMapper.Normalize));

        private static readonly MethodInfo _normalizeNullableMethod =
            typeof(DateTimeKindMapper).GetTypeInfo().GetMethod(nameof(DateTimeKindMapper.NormalizeNullable));

        private static readonly MethodInfo _normalizeObjectMethod =
            typeof(DateTimeKindMapper).GetTypeInfo().GetMethod(nameof(DateTimeKindMapper.NormalizeObject));

        public override Expression CreateReadValueExpression(Expression valueBuffer, Type type, int index, IProperty property = null)
        {
            if (type == typeof(DateTime))
            {
                return Expression.Call(
                    _normalizeMethod,
                    base.CreateReadValueExpression(valueBuffer, type, index, property));
            }

            if (type == typeof(DateTime?))
            {
                return Expression.Call(
                    _normalizeNullableMethod,
                    base.CreateReadValueExpression(valueBuffer, type, index, property));
            }

            return base.CreateReadValueExpression(valueBuffer, type, index, property);
        }

        public override Expression CreateReadValueCallExpression(Expression valueBuffer, int index)
        {
            var readValueCallExpression = base.CreateReadValueCallExpression(valueBuffer, index);
            if (readValueCallExpression.Type == typeof(DateTime))
            {
                return Expression.Call(
                    _normalizeMethod,
                    readValueCallExpression);
            }

            if (readValueCallExpression.Type == typeof(DateTime?))
            {
                return Expression.Call(
                    _normalizeNullableMethod,
                    readValueCallExpression);
            }

            if (readValueCallExpression.Type == typeof(object))
            {
                return Expression.Call(
                    _normalizeObjectMethod,
                    readValueCallExpression);
            }

            return readValueCallExpression;
        }
    }
```

Workaround from volosoft works, except cases in subquery, like:

```c#
context.Entity.Select(e => new
{
    DateTimeField = e.NavigationProperty.OrderBy(n => n.OrderField).FirstOrDefault().DateTimeField
});
```

So extended `EntityMaterializerSource` can handle this as well, but because it wraps every `object` response as well, it can cause additional performance cost because of `object` to `DateTime` cast.

And it has additional fix for nullable `DateTime` as well. Before 2.1 we will use this workaround.

**Comment by EEaglehouse** [comment]
Hi, Alexander,

That looks like a nice solution and more advanced than the solution we have been using, which was to use T4 to generate a custom DBContext. The code you presented is pretty easy to understand and not very complicated. When we migrate to .NET Core, this will definitely come in handy.

Thank you very much for taking the time to share it with me!

--Ed

**Comment by ajcvickers** [comment]
Please note that EntityMaterializerSource is internal code and any solution like that shown above may get broken with a new release of EF. That doesn't mean I'm saying not to use it--it's publicly available internal code for a reason--but please be aware of the risk associated with it.

Starting in EF Core 2.1, this would be one way to deal with DateTime.Kind:
```C#
modelBuilder
    .Entity<Foo>()
    .Property(e => e.SomeDate)
    .HasConversion(v => v, v => DateTime.SpecifyKind(v, DateTimeKind.Local));
```
This will ensure every time the date is read from the database it is specified as Local automatically. It will also work in projections, etc. which the code above does not handle.

**Comment by werwolfby** [comment]
@ajcvickers but my snippet above works with projections as well.

**Comment by ajcvickers** [comment]
@werwolfby You are correct--I copy pasted that code snippet from another post and missed that last sentence, which was relevant there but not here. The workaround you posted is a good one--this is probably the best way to make this work, and is, in effect, what the value converter in 2.1 is doing.

Being able to do things like this is why we made the internal code available. However, we have to be careful that internal code is only used when people understand the risks. Otherwise too many people may start complaining when they get broken and we will be under pressure to make it really internal again.

**Comment by werwolfby** [comment]
@ajcvickers Yes I understand the risk using this code. And I completely agree, but we have to use it, and thank you to make this API public available 😄.
And I create gist to keep this "helper" up-to-date until EF Core 2.1: https://gist.github.com/werwolfby/7f04558bc21c8114e209d5727fb2e9f8

**Comment by werwolfby** [comment]
@martavoi This is expected, because you're selecting anonymous object, and you didn't (and can't) specify conversion for anonymous type. And even if `Created` and `Estimated` has conversion, this will not help because it was converted to Sql without any conversion.

And as far as I understand you can't solve this issue with current `.HasConversion` functionality.
Because you can specify it only for Entities, but not for anonymous type.

There should be something to map to views, maybe it will help.

**Comment by martavoi** [comment]
@werwolfby i would say there must be smth allow us to configure DateTime materialization for entire Context. We usually store UTC dates and expect to see DateTime.Kind == UTC for all queries

**Comment by werwolfby** [comment]
@martavoi, we had tried to migrate to EF Core 2.1 as well. And yep, only few queries starts to work, but all this projection still doesn't work. This is because all this Value Converter can be set for properties only, its not affect projection. So we still continue to use my workaround from above 😞  ([gist](https://gist.github.com/werwolfby/7f04558bc21c8114e209d5727fb2e9f8))

We still need something from EF, to be able work on projections as well. CCed @ajcvickers.

For example this projection works partially, but I expected that it should works always:

```c#
context.Entities.Select(e => new ProjectionEntity
{
   LastTested = e.Results
        .Where(x => x.CompletedDate.HasValue)
        .OrderByDescending(x => x.CompletedDate)
        .FirstOrDefault().
        .CompletedDate,
    RevisionDate = e.CreationDate
}
```
And `RevisionDate` converted right using value converter, while `CompletedDate` don't.
Both properties had converters.

I think this issue should be reopened.

**Comment by ajcvickers** [comment]
@martavoi Can you please post a new issue with a complete runnable project/solution or code listing that reproduces the behavior you are seeing? It would be good to infer the converter to use for Deadline, but I don't know how feasible it is now.

@werwolfby Can you also post another new issue with a complete runnable project/solution that reproduces what you are seeing? I think we need to understand better how that query is being executed in order to understand what is going on.

**Comment by werwolfby** [comment]
@ajcvickers Will try to implement it in few days.

But the main question here is still the same. Do you plan to allow specify converters globally on type, instead of on each property?

**Comment by ajcvickers** [comment]
@werwolfby This is tracked by #10784

**Comment by springy76** [comment]
The EF 6.x provider of System.Data.SQLite has an option for the connection string which lets you configure which DateTimeKind *ANY* DateTime retrieved by the provider will have no matter if entities or anonymous objects or whatsoever. *SINCE YEARS!*

**Comment by forlayo** [comment]
I am using the solution that @rowanmiller suggest; and I would say that it should be considered as correct solution. 

Because, as he points, SQL Server does not allow to store that datetime is on an specific timezone, and setting a default one break the consistency conceptually through the system.

You can do same you're doing with datetime with datetimeoffset and it would include the timezone, so why not using it?

**Comment by Compufreak345** [comment]
If anyone finds it helpful - I modified the great snippet from @ajcvickers to make sure all my dates are saved as UTC and read as UTC.

```
foreach (var entityType in modelBuilder.Model.GetEntityTypes())
            {
                    foreach (var property in entityType.GetProperties())
                    {
                        if (property.ClrType == typeof(DateTime))
                        {
                            modelBuilder.Entity(entityType.ClrType)
                                .Property<DateTime>(property.Name)
                                .HasConversion(
                                    v => v.ToUniversalTime(),
                                    v => DateTime.SpecifyKind(v, DateTimeKind.Utc));
                        }

                        if (property.ClrType == typeof(DateTime?))
                        {
                            modelBuilder.Entity(entityType.ClrType)
                                .Property<DateTime?>(property.Name)
                                .HasConversion(
                                    v => v.HasValue ? v.Value.ToUniversalTime() : v,
                                    v => v.HasValue ? DateTime.SpecifyKind(v.Value, DateTimeKind.Utc) : v);
                        }
                    }
            }
```

**Comment by springy76** [comment]
@Compufreak345 will this work with queries composing into anonymous (or non-entity) types?

**Comment by Compufreak345** [comment]
> @Compufreak345 will this work with queries composing into anonymous (or non-entity) types?

I am not sure what you mean by "queries composing into anonymous types", if you do your writing database transactions using LINQ to SQL with entities it should work for everything you do. It should also work if you convert your properties to another object type while reading LINQ entities, if you do something outside of entity framework like hard-coded SQL queries it will not.

**Comment by clement-fifty** [comment]
Note that in my case, I have queries (View equivalent) and that doesn't seem to like this workaround, so I had to use the following :

```
foreach(var entityType in builder.Model.GetEntityTypes()) {
    if (entityType.IsQueryType) continue; // This is the difference!

    foreach(var property in entityType.GetProperties()) {
        if (property.ClrType == typeof(DateTime)) {
            builder.Entity(entityType.ClrType)
                .Property < DateTime > (property.Name)
                .HasConversion(
                    v => v.ToUniversalTime(),
                    v => DateTime.SpecifyKind(v, DateTimeKind.Utc));
        } else if (property.ClrType == typeof(DateTime ? )) {
            builder.Entity(entityType.ClrType)
                .Property < DateTime ? > (property.Name)
                .HasConversion(
                    v => v.HasValue ? v.Value.ToUniversalTime() : v,
                    v => v.HasValue ? DateTime.SpecifyKind(v.Value, DateTimeKind.Utc) : v);
        }
    }
}
```

**Comment by ChristopherHaws** [comment]
@clement-fifty I wasn't able to get your solution working whenever I had the property manually configured (for example if I give the column a custom name in the db). I was able to make it work by doing the following though:

```c#
var dateTimeConverter = new ValueConverter<DateTime, DateTime>(
    v => v.ToUniversalTime(),
    v => DateTime.SpecifyKind(v, DateTimeKind.Utc));

var nullableDateTimeConverter = new ValueConverter<DateTime?, DateTime?>(
    v => v.HasValue ? v.Value.ToUniversalTime() : v,
    v => v.HasValue ? DateTime.SpecifyKind(v.Value, DateTimeKind.Utc) : v);

foreach (var entityType in builder.Model.GetEntityTypes())
{
    if (entityType.IsQueryType)
    {
        continue;
    }

    foreach (var property in entityType.GetProperties())
    {
        if (property.ClrType == typeof(DateTime))
        {
            property.SetValueConverter(dateTimeConverter);
        }
        else if (property.ClrType == typeof(DateTime?))
        {
            property.SetValueConverter(nullableDateTimeConverter);
        }
    }
}
```

**Comment by btecu** [comment]
`.IsQueryType` doesn't seem to be present for EF Core 3.

**Comment by smitpatel** [comment]
`entityType.IsKeyless`

**Comment by jeremycook** [comment]
I also check the column type since I have some 'date' fields in my database that are not UTC. For a SQL Server datatime field that looks like this:

```csharp
property.GetColumnType() == "datetime"
```

**Comment by ChristopherHaws** [comment]
@jeremycook I recently have run into a similar problem where certain fields I didnt want to be tracked as UTC, so I made the following changes to my code:

```csharp
public class ApplicationContext : DbContext
{
    protected override void OnModelCreating(ModelBuilder builder)
    {
        base.OnModelCreating(builder);

        builder.ApplyConfigurationsFromAssembly(Assembly.GetExecutingAssembly());

        var dateTimeConverter = new ValueConverter<DateTime, DateTime>(v => v, v => DateTime.SpecifyKind(v, DateTimeKind.Utc));
        var nullableDateTimeConverter = new ValueConverter<DateTime?, DateTime?>(v => v, v => v.HasValue ? DateTime.SpecifyKind(v.Value, DateTimeKind.Utc) : v);

        foreach (var entityType in builder.Model.GetEntityTypes())
        {
            foreach (var property in entityType.GetProperties())
            {
                if (!property.IsUtc())
                {
                    continue;
                }

                if (property.ClrType == typeof(DateTime))
                {
                    property.SetValueConverter(dateTimeConverter);
                }
                else if (property.ClrType == typeof(DateTime?))
                {
                    property.SetValueConverter(nullableDateTimeConverter);
                }
            }
        }
    }
}

public static class UtcDateAnnotation
{
    private const String IsUtcAnnotation = "IsUtc";
    public static PropertyBuilder<TProperty> IsUtc<TProperty>(this PropertyBuilder<TProperty> builder, Boolean isUtc = true) =>
        builder.HasAnnotation(IsUtcAnnotation, isUtc);

    public static Boolean IsUtc(this IMutableProperty property) =>
        ((Boolean?)property.FindAnnotation(IsUtcAnnotation)?.Value) ?? true;
}
```

Which allows me to mark a property as not UTC with the following code (UTC is true by default):
```csharp
builder.Property(x => x.DateField).IsUtc(false);
```

**Comment by JonPSmith** [comment]
Can I ask a question on nullable value converters. The EF docs says

> A null value will never be passed to a value converter. This makes the implementation of conversions easier and **allows them to be shared amongst nullable and non-nullable properties**.

(I bolded the part in question).

But [The comment by @ChristopherHaws](https://github.com/dotnet/efcore/issues/4711#issuecomment-535288442) shows two versions for a `DateTime` and `DateTime?`.  Is that necessary or can you assume that your converter won't be called if the value is null?

**Comment by EEaglehouse** [comment]
Great question. The way I read that note, if the only difference in the second ValueConverter is checking for null, it won't be called anyway. In the DateTime? case here, it would be unnecessary.

But thinking about it, it says a null value will never be passed *to* a value converter and this *allows* them to be shared. But it does not say a null value cannot be returned *from* a value converter (say for example, converting an empty string to a null), so a nullable option could still be useful.

Would somebody who has experience with this please comment?

**Comment by ajcvickers** [comment]
@JonPSmith @EEaglehouse Null will never be passed to a converter, so the same non-nullable converter can be used for the PK and the FK. We propagate the converter set on the PK automatically to FKs, so you only really need to set the converter on the PK.

Returning null from a converter is possible, but may result in unexpected behavior for FKs since it changes the relationship.

**Comment by ChristopherHaws** [comment]
@EEaglehouse I updated the code snippet to exclude the nullable case. Thanks @ajcvickers for confirming that. 👍

**Comment by gabynevada** [comment]
@ChristopherHaws I modified the code snippet to include Utc conversion for nullable DateTime objects (DateTime?)
```c#
public class ApplicationContext : DbContext
{
  protected override void OnModelCreating(ModelBuilder builder)
  {
      base.OnModelCreating(builder);

      builder.ApplyConfigurationsFromAssembly(Assembly.GetExecutingAssembly());
      builder.ApplyUtcDateTimeConverter();
  }
}

public static class UtcDateAnnotation
{
  private const String IsUtcAnnotation = "IsUtc";
  private static readonly ValueConverter<DateTime, DateTime> UtcConverter =
    new ValueConverter<DateTime, DateTime>(v => v, v => DateTime.SpecifyKind(v, DateTimeKind.Utc));

  private static readonly ValueConverter<DateTime?, DateTime?> UtcNullableConverter =
    new ValueConverter<DateTime?, DateTime?>(v => v, v => v == null ? v : DateTime.SpecifyKind(v.Value, DateTimeKind.Utc));

  public static PropertyBuilder<TProperty> IsUtc<TProperty>(this PropertyBuilder<TProperty> builder, Boolean isUtc = true) =>
    builder.HasAnnotation(IsUtcAnnotation, isUtc);

  public static Boolean IsUtc(this IMutableProperty property) =>
    ((Boolean?)property.FindAnnotation(IsUtcAnnotation)?.Value) ?? true;

  /// <summary>
  /// Make sure this is called after configuring all your entities.
  /// </summary>
  public static void ApplyUtcDateTimeConverter(this ModelBuilder builder)
  {
    foreach (var entityType in builder.Model.GetEntityTypes())
    {
      foreach (var property in entityType.GetProperties())
      {
        if (!property.IsUtc())
        {
          continue;
        }

        if (property.ClrType == typeof(DateTime))
        {
          property.SetValueConverter(UtcConverter);
        }

        if (property.ClrType == typeof(DateTime?))
        {
          property.SetValueConverter(UtcNullableConverter);
        }
      }
    }
  }
}
```

**Comment by ChristopherHaws** [comment]
> @JonPSmith @EEaglehouse Null will never be passed to a converter, so the same non-nullable converter can be used for the PK and the FK. We propagate the converter set on the PK automatically to FKs, so you only really need to set the converter on the PK.
> 
> Returning null from a converter is possible, but may result in unexpected behavior for FKs since it changes the relationship.

@gabynevada Per this comment, you shouldn't need to handle null, that is why I removed it from my snippet. Converters don't get called for null values.

**Comment by rallets** [comment]
@ChristopherHaws I can confirm that if I have a nullable DateTime I need an explicit DateTime? converter as posted by @gabynevada

**Comment by JonPSmith** [comment]
Hi @rallets,

Doesn't look like @ChristopherHaws is around. I can confirm that you don't need an explicit DateTime? converter. A null value (of any type) is just passed through as a null would calling your Value Converter. See [this note at the in the EF Core docs on Value Converters](https://docs.microsoft.com/en-us/ef/core/modeling/value-conversions#configuring-a-value-converter).

I hope that helps.

**Comment by jeremycook** [comment]
@JonPSmith my experience matches @rallets ... Maybe there is some confusion over configuration for a nullable type vs. a null value being passed. I found that configuration is needed for both the nullable and non-nullable version of a DateTime if your model has properties of both types.

To clarify by example (and this was with EF Core 3 or 3.1, not sure if 5.0 behaves different) if I have a property like `public DateTime? Created { get; set; }` and I want to apply a type converter to it, the signature must match `DateTime?` and not simply `DateTime`. Perhaps this is a bug or somehow we are not using the type converter configuration API correctly!?

**Comment by rallets** [comment]
@JonPSmith Thanks, I see where the difference is. 
The link you are providing is about the built-in `modelBuilder.[...].HasConversion(converter)`.
In the snippet used to apply ~globally~ automatically the validators, there is a conditional check on the property type:
`if (property.ClrType == typeof(DateTime))`
this means that if the type of the property is `DateTime?` the converter is not applied.

Hence in this case we need to handle the DateTime?, too:
```
if (property.ClrType == typeof(DateTime?)) {
	property.SetValueConverter(UtcNullableConverter);
}
```
Probably it's possible to make it working using only one converter (UtcConverter) for both types:
`if (property.ClrType == typeof(DateTime) || property.ClrType == typeof(DateTime?))`
but I need to test it for confirmation.

**Comment by JonPSmith** [comment]
Yes @rallets, I didn't realise that you were referring to @gabynevada example which applies a value converter. Yes, changing the test would fix that. 

Also see [Arthur Vickers' comment](https://github.com/dotnet/efcore/issues/4711#issuecomment-642681865) in this issue. If anyone knows the right answer, then Arthur does!

**Comment by joakimriedel** [comment]
> Probably it's possible to make it working using only one converter (UtcConverter) for both types:
> `if (property.ClrType == typeof(DateTime) || property.ClrType == typeof(DateTime?))`
> but I need to test it for confirmation.

Just verified that this works in my codebase, using the same value converter for both nullable and non-nullable. Thanks for the tip @rallets !

**Comment by ChristopherHaws** [comment]
Thanks for confirming that bug in my snippet. I have updated the snippet to reflect the changes. 👍

**Comment by peabnuts123** [comment]
Is it possible to wrap up some of the code-snippets posted above into some kind of annotation? Like
```csharp
  [Required]
  [HasDateTimeKind(DateTimeKind.Utc)]
  public DateTime CreatedOn { get; set; }
```

My knowledge of customising EFCore's serialisation etc is only limited. If possible, maybe this could even land in something like `System.ComponentModel.Annotations`? Seems like there is a pretty wide use-case/need for people specifying the kind, as database providers to not encode this information.

**Comment by jeremycook** [comment]
@peabnuts123 That would be convenient. That attribute and the annotations folks graciously provided here still don't overcome the fact that you can still shoot yourself in the foot with DateTime...like when you instantiate a DateTime object without specifying its kind, or by specifying the wrong kind like when DateTime.Now is used when DateTime.UtcNow should have been. For folks who are growing tired of worrying about DateTime I suggest just using DateTimeOffset for now (no pun intended) and/or upvoting the [Native support for NodaTime issue](https://github.com/dotnet/efcore/issues/20114).

**Comment by ChristopherHaws** [comment]
@peabnuts123 Here is an updated version of my snippet that includes an attribute called `IsUtc` the behaves the same as the builder method:

```csharp
public static class UtcDateAnnotation {
    private const string IsUtcAnnotation = "IsUtc";
    private static readonly ValueConverter<DateTime, DateTime> UtcConverter =
        new ValueConverter<DateTime, DateTime>(v => v, v => DateTime.SpecifyKind(v, DateTimeKind.Utc));

    public static PropertyBuilder<TProperty> IsUtc<TProperty>(this PropertyBuilder<TProperty> builder, bool isUtc = true) =>
        builder.HasAnnotation(IsUtcAnnotation, isUtc);

    public static bool IsUtc(this IMutableProperty property) {
        var attribute = property.PropertyInfo.GetCustomAttribute<IsUtcAttribute>();
        if (attribute is not null && attribute.IsUtc) {
            return true;
        }
        
        return ((bool?)property.FindAnnotation(IsUtcAnnotation)?.Value) ?? true;
    }

    /// <summary>
    /// Make sure this is called after configuring all your entities.
    /// </summary>
    public static void ApplyUtcDateTimeConverter(this ModelBuilder builder) {
        foreach (var entityType in builder.Model.GetEntityTypes()) {
            foreach (var property in entityType.GetProperties()) {
                if (!property.IsUtc()) {
                    continue;
                }


                if (property.ClrType == typeof(DateTime) ||
                    property.ClrType == typeof(DateTime?)) {
                    property.SetValueConverter(UtcConverter);
                }
            }
        }
    }
}

public class IsUtcAttribute : Attribute {
    public IsUtcAttribute(bool isUtc = true) => this.IsUtc = isUtc;
    public bool IsUtc { get; }
}
```

Something to note though, this snippet is not intended to "convert" the datetime into utc, but rather is used to specify the kind when reading the value from the database. You still need to be diligent and only ever pass in the correct type of DateTime. As far as I know it is not possible to write a converter that is smart enough to do the correct conversion logic since expressions are pretty limited. If it were possible though, DateTimeKind.Unspecified would probably be unhandled and should throw an exception since there is no way to know what timezone it is in.

The built in DateTime functionality is weird (hench why NodaTime exists). The build in functionality makes a lot of assumptions that could (and often are) wrong:
- DateTimeKind.Utc
    - ToUtc does nothing
    - ToLocalTime converts the datetime to the system's timezone
- DateTimeKind.Local
    - ToUtc converts the datetime to UTC timezone
    - ToLocalTime does nothing
- DateTimeKind.Unspecified (this is where things get weird)
    - ToUtc assumes the datetime is in the systems timezone and converts to UTC  from the system timezone
    - ToLocalTime assumes the datetime is in UTC and converts to the systems timezone from UTC

**Comment by peabnuts123** [comment]
That's top notch @ChristopherHaws. Thanks for the code and your dedication to this issue. That gets pretty much as close to an optimal solution as I think you can achieve without EF Core adopting it natively.

**Comment by acbaile** [comment]
Offered is the best solution. Attribute or fluent API call. It will be convenient. Time in DB can be stored by the most economic way, but when EF Core takes it from DB and puts into C# property, it creates DateTime with assigned Kind. If it is not yet, it must be added to future functionality of EF Core.

[HasDateTimeKind(DateTimeKind.Utc)]
public DateTime CreatedOn { get; set; }

**Comment by ChristopherHaws** [comment]
@acbaile I am also storing all my dates as UTC in the database. The problem occurs when you read the dates out of the database. The default behavior is to read the values as "unspecified", however the dates are actually in UTC. This means that if I we to call `.ToUniversalTime()` on them, I would now end up with the incorrect date. This gets particularly interesting when you further serialize the values into json or xml (as is my use case since the values are returned from an  ASP.NET Core WebAPI). In this scenario the values would be read from the database as unspecified and then serialized to JSON without a 'Z' at the end to specify that it is zulu (UTC) time which would in turn be translated incorrectly by front-end libraries such as momentjs. Instead of constantly having to specify the kind on every datetime I read from the database, I wrote that snippet to do the work for me. :)

Here is a short snippet to show the problem:
```csharp
void Main() {
    var model = new {
        UtcDateTime = DateTime.UtcNow,
        UnspecifiedDateTime = DateTime.SpecifyKind(DateTime.UtcNow, DateTimeKind.Unspecified)
    };

    var json = JsonConvert.SerializeObject(new {
        UtcDateTime = DateTime.UtcNow,
        UnspecifiedDateTime = DateTime.SpecifyKind(DateTime.UtcNow, DateTimeKind.Unspecified),
        UnspecifiedDateTimeOffset = new DateTimeOffset(DateTime.SpecifyKind(DateTime.UtcNow, DateTimeKind.Unspecified))
    }, Newtonsoft.Json.Formatting.Indented);
    
    Console.WriteLine(json);

    /*
    Outputs:
    {
      "UtcDateTime": "2020-11-30T09:00:32.1302999Z",
      "UnspecifiedDateTime": "2020-11-30T09:00:32.1303004",
      "UnspecifiedDateTimeOffset": "2020-11-30T09:00:32.1303008-08:00"
    }
    */
}

```

Hope that helps clarify!

**Comment by mike-whitehurst-b2m** [comment]
@ChristopherHaws  Was there a reason for IsQueryType/IsKeyLess being in the earlier code sample? It results in no "Z" when reading from a View. Checking it's safe to remove.

**Comment by ChristopherHaws** [comment]
@mike-whitehurst-b2m I don't remember the exact reason since I have since removed that logic. I do know that the EF setup I use is slightly unconventional (I use mostly owned types using SQL provider) and there might have been a bug at the time that caused me to need that logic (I have found and reported several bugs related to owned types since I use them extensively). I would recommend you try out [this](https://github.com/dotnet/efcore/issues/4711#issuecomment-734471122) snipped and make sure you add a few unit and integration tests that test your specific scenarios. Perhaps in the future I will try to rewrite the logic to be interceptors and package it as a library, I was just hoping we would get support for this in EF Core directly. :)

**Comment by fartwhif** [comment]
five years and this bites people in the ass still, but yeah, "closed" give me a break.

**Comment by roji** [comment]
@fartwhif EF Core already allows you to specify the DateTime.Kind coming back from the databases - via value converters - and 6.0 even allows you to configure that globally for all DateTime properties in your model. Beyond that it's hard to see what EF Core could do - SQL Server's datetime/datetime2 types simply do not store DateTimeKind, which is a .NET-only concept. Are you looking for anything in particular?

**Comment by ChristopherHaws** [comment]
@roji IMO, the only real solution is to store the value in two columns (datetime and timezone name). Unfortunately DateTimeOffset is not a real solution for this problem since many timezone's exist within a single offset (for example Arizona and Colorado are both in Mountain Time however Arizona does not have daylight savings while Colorado does) and those timezones could have different rules about daylight savings time. Since this isn't a feature of EF Core, I always store all my datetime's in UTC and manually store the timezone name when necessary. If only daylight savings didn't exist :P

**Comment by roji** [comment]
> the only real solution is to store the value in two columns (datetime and timezone name)

Note that there's a very big difference between storing the DateTime's timestamp and **kind**, and storing a timestamp and timezone - this issue really is about fully persisting a DateTime, which does not have a timezone name (or even an offset). In other words, it's not just the database that doesn't have a single type for timestamp+timezone, it's also .NET (I recommend looking at NodaTime, which does have [ZonedDateTIme](https://nodatime.org/3.0.x/api/NodaTime.ZonedDateTime.html)). Other than that, if your goal is to save a timestamp and timezone name, then you're right that you need two columns ([see this blog post](https://www.roji.org/storing-timezones-in-the-db) which I just recently wrote about the subject).

To summarize, on the .NET side you need two properties (DateTime and string, for the time zone), and on the SQL Server side you need two columns (datetime2 and nvarchar). If you want to use NodaTime (highly recommended), then on the .NET side you can have a single property of time ZonedDateTime. At the moment EF Core cannot map a single .NET property to multiple columns, but that feature is planned for EF7 (#13947).

**Comment by roji** [comment]
@SebFerraro above I just referred to the ability to configure value converters to modify the Kind on DateTimes being read, [as here](https://docs.microsoft.com/en-us/ef/core/modeling/value-conversions?tabs=data-annotations#specify-the-datetimekind-when-reading-dates). That can be combined with the new EF Core 6.0 [bulk configuration](https://docs.microsoft.com/en-us/ef/core/modeling/bulk-configuration#pre-convention-configuration) to apply this globally to your model.

But I want to stress again that this isn't the ideal way to do it. If what you want is to simply store UTC timestamps and to retrieve them as UTC DateTimes, use version 6.0 and map your EF DateTime properties to `timestamp with time zone`. This will automatically make DateTimes coming back have UTC Kind without any value converters.

**Comment by joakimriedel** [comment]
@SebFerraro the bulk configuration using `ConfigureConventions` does not work with the [new compiled models in EF Core 6](https://devblogs.microsoft.com/dotnet/announcing-entity-framework-core-6-0-preview-5-compiled-models/), so we decided to adapt the code from @ChristopherHaws above slightly. Works like a charm!

```c#
using System;
using System.Reflection;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata;
using Microsoft.EntityFrameworkCore.Metadata.Builders;
using Microsoft.EntityFrameworkCore.Storage.ValueConversion;

namespace Whatever;

    public class UtcValueConverter : ValueConverter<DateTime, DateTime>
    {
        public UtcValueConverter()
            : base(v => v, v => DateTime.SpecifyKind(v, DateTimeKind.Utc))
        {
        }
    }

    public static class UtcDateAnnotation
    {
        private const string IsUtcAnnotation = "IsUtc";

        public static PropertyBuilder<TProperty> IsUtc<TProperty>(this PropertyBuilder<TProperty> builder, bool isUtc = true)
        {
            if (builder is null)
            {
                throw new ArgumentNullException(nameof(builder));
            }

            return builder.HasAnnotation(IsUtcAnnotation, isUtc);
        }

        public static bool IsUtc(this IMutableProperty property)
        {
            if (property is null)
            {
                throw new ArgumentNullException(nameof(property));
            }

            var attribute = property.PropertyInfo?.GetCustomAttribute<IsUtcAttribute>();
            if (attribute is not null && attribute.IsUtc)
            {
                return true;
            }

            return ((bool?)property.FindAnnotation(IsUtcAnnotation)?.Value) ?? true;
        }

        /// <summary>
        /// Make sure this is called after configuring all your entities.
        /// </summary>
        public static void ApplyUtcDateTimeConverter(this ModelBuilder builder)
        {
            if (builder is null)
            {
                throw new ArgumentNullException(nameof(builder));
            }

            foreach (var entityType in builder.Model.GetEntityTypes())
            {
                foreach (var property in entityType.GetProperties())
                {
                    if (!property.IsUtc())
                    {
                        continue;
                    }

                    if (property.ClrType == typeof(DateTime) ||
                        property.ClrType == typeof(DateTime?))
                    {
                        property.SetValueConverter(typeof(UtcValueConverter));
                    }
                }
            }
        }
    }

    [AttributeUsage(AttributeTargets.Property)]
    public class IsUtcAttribute : Attribute
    {
        public IsUtcAttribute(bool isUtc = true) => this.IsUtc = isUtc;
        public bool IsUtc { get; }
    }
```

then in **THE END OF** (important!) `OnModelCreating` just put

```
            // all entity configuration

            // note: this line has to be last in this method to work!

            builder.ApplyUtcDateTimeConverter();
```

so your developers only need to remember never put any configuration below that line, and all properties will be automatically converted for `DateTimeKind.Utc` on read.

@roji slightly off topic: the links in the first paragraph of the docs you sent a link for are broken (see "xref:") https://docs.microsoft.com/en-us/ef/core/modeling/bulk-configuration#pre-convention-configuration

**Comment by SebFerraro** [comment]
@joakimriedel it was an existing database that I ported into a brand new ef core project, and your solution has actually worked, which is great! Thanks!

**Comment by roji** [comment]
@joakimriedel @SebFerraro if what you want is for DateTime to always come back as UTC, there's no need for an attribute; @SebFerraro your code didn't work since HaveConversion in ConfigureConventions only accepts a value converter type, rather than lambdas (in order to better support compiled models). See below for a minimal code sample which shows everything working without an attribute.

<details>
<summary>Code sample</summary>

```c#
await using var ctx = new BlogContext();
await ctx.Database.EnsureDeletedAsync();
await ctx.Database.EnsureCreatedAsync();

var blog = await ctx.Blogs.SingleAsync();
Console.WriteLine(blog.DateTime.Kind);

public class BlogContext : DbContext
{
    public DbSet<Blog> Blogs { get; set; }

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        => optionsBuilder
            .UseSqlServer(@"Server=localhost;Database=test;User=SA;Password=Abcd5678;Connect Timeout=60;ConnectRetryCount=0")
            .LogTo(Console.WriteLine, LogLevel.Information)
            .EnableSensitiveDataLogging();

    protected override void OnModelCreating(ModelBuilder modelBuilder)
        => modelBuilder.Entity<Blog>().HasData(new Blog { Id = 1, DateTime = DateTime.Now });

    protected override void ConfigureConventions(ModelConfigurationBuilder configurationBuilder)
    {
        configurationBuilder
            .Properties<DateTime>()
            .HaveConversion(typeof(UtcValueConverter));
    }

    class UtcValueConverter : ValueConverter<DateTime, DateTime>
    {
        public UtcValueConverter()
            : base(v => v, v => DateTime.SpecifyKind(v, DateTimeKind.Utc))
        {
        }
    }
}

public class Blog
{
    public int Id { get; set; }
    public DateTime DateTime { get; set; }
}
```

</details>

> @roji slightly off topic: the links in the first paragraph of the docs you sent a link for are broken (see "xref:") https://docs.microsoft.com/en-us/ef/core/modeling/bulk-configuration#pre-convention-configuration

Thanks. The 6.0 API docs (which this links point to) have only recently been pushed online in https://github.com/dotnet/EntityFramework.Docs/issues/3630, and I don't think we've pushed our conceptual docs since then - this should go away the next time we do.

**Comment by joakimriedel** [comment]
@roji the attribute is a negative option, so the code will actually set the value converter for all DateTime properties without using any attributes.

But regarding the `ConfigureConventions`, I actually never tried it since the method has the following remark:

```
If a model is explicitly set on the options for this context (via Microsoft.EntityFrameworkCore.DbContextOptionsBuilder.UseModel(Microsoft.EntityFrameworkCore.Metadata.IModel)) then this method will not be run.
```

Since I expicitly set a model via `UseModel `I never even bothered to try it.

But you're saying it should work with compiled models? I will try right away! 

Perhaps then the wording of this remark should be revisited?

**Comment by qsdfplkj** [comment]
In order to save it as Utc I use:

    builder.Property(x => x.DateTime).HasConversion((x) => x.ToUniversalTime(), (x) => DateTime.SpecifyKind(x, DateTimeKind.Utc) );

This way if someone uses unspecified or local time at least it is converted to utc time.

**Comment by ChristopherHaws** [comment]
@qsdfplkj Be carful with this approach. If the DateTimeKind is Unspecified and you call ToUniversalTime, it will assume that the DateTime is in local time which can cause some hard to find bugs.

**Comment by qsdfplkj** [comment]
Thanks I will check this serialization scenario.

**Comment by palhal** [comment]
Here's my take on this for a nullable DateTime, inspired by @qsdfplkj's and @ChristopherHaws' answers: 

```csharp
public DateTime? TimestampUtc { get; set; }
```
```csharp
builder.Property(x => x.TimestampUtc)
    .HasConversion(new UtcDateTimeConverter());
```
```csharp
public class UtcDateTimeConverter : ValueConverter<DateTime?, DateTime?>
{
    public UtcDateTimeConverter()
        : base(
            d => !d.HasValue ? null : ConvertToUtc(d.Value),
            d => !d.HasValue ? null : SpecifyUtc(d.Value))
    {
    }

    private static DateTime ConvertToUtc(DateTime date)
    {
        switch (date.Kind)
        {
        case DateTimeKind.Utc:
            return date;
        case DateTimeKind.Local:
            return date.ToUniversalTime();
        default:
            throw new InvalidTimeZoneException($"Unsupported DateTimeKind: {date.Kind}");
        }
    }

    private static DateTime SpecifyUtc(DateTime date)
    {
        return DateTime.SpecifyKind(date, DateTimeKind.Utc);
    }
}
```

**Comment by qsdfplkj** [comment]
I didn't use this approach anymore because usually data comes in from an API and it is unspecified. So I decided that unspecified is the actually the better way. (Maybe use DateTimeOffset?)

**Comment by palhal** [comment]
Thanks for the suggestion. I actually use both types. My model looks like this:
```csharp
public DateTimeOffset Timestamp { get; set; }
public DateTime InsertedAtUtc { get; set; }
```

Timestamp is when an event actually occured. It's useful for our users to know the offset. (Yes, the API requires an offset to be included.) On the other hand, the insertion date is automatically set in the backend, and only UTC makes sense here. Maybe I'm overthinking, but I think that having the possibility to store an offset creates ambiguity.

**Comment by springy76** [comment]
BTW: npgsql (postgres) provider just disallows (per default) any non-utc DateTime, both for storage and query operations. And always return uct, too - of course.

**Comment by roji** [comment]
@qsdfplkj 

> I didn't use this approach anymore because usually data comes in from an API and it is unspecified.

Carefully consider what exactly this means: are you really saying that you have no idea what time zone the timestamps are, which your API provides? If that's true, you can't really do anything with that data, since it could refer to arbitrary moments in time, etc. It's more common for the time zone to be simply implicit, e.g. for the timestamps to be in UTC or some very specific local timestamp - even if it's not specified. If that's the case, I highly recommend changing the DateTime's Kind to Utc (via DateTime.SpecifyKind), right where you're reading the timestamps, and not when writing them to the database (if you're deserializing JSON from the API, you can probably configure the JSON deserializer to do this for you as well).

Note that DateTimeOffset also implies that you know which time zone the timestamps are in.

**Comment by qsdfplkj** [comment]
I guess I've to revisit this but timestamp is always Utc. But what is a local times on a server in azure anyway?

**Comment by ericbl** [comment]
after reading all this discussion and the related [npgsql doc](https://www.npgsql.org/efcore/release-notes/6.0.html#migrating-columns-from-timestamp-to-timestamptz), I am still a bit confused about what I should do in our very classic use case:

- we are developing a service with web api and data exchange with mqtt. So input/output with System.Text.JsonSerializer
- DB postgres using npgsql with EF Core 6.0.25
- we want to use only UTC time
- we have plenty of C# code set with DateTime and not so motivated to switch to DateTimeOffset
- we have some Entities with DateTime 

`add-migration Initial`  properly create the DateTime (type: "timestamp with time zone" )
and the seed method with DateTimeKind.Utc

 dotnet swagger generate the openapi.json with properties with `"format": "date-time"`

JsonSerializer also defined  UTC per default AKAIK .

Everything seem to work fine in our first tests.

So apparently, there is NOTHING else to add, no need for DateTimeOffset, special converter and so one if UTC is the default all the way.
However, the whole documentation is not very clear there...

**Comment by roji** [comment]
@ericbl the details vary somewhat across databases. Generally speaking, there's no need at all to switch to DateTimeOffset for using UTC timestamps; a DateTime with Kind=Utc will do just as well. However, since e.g. SQL Server doesn't store the Kind, DateTime instances read from it will have Kind=Unspecified; this could be a reason to use DateTimeOffset instead, since in that case an offset of 0 can be stored in the database. Or you can use an EF value converter to ensure that DateTimes read from the database have Kind=Utc. In PostgreSQL that particular problem doesn't exist.

Beyond that, if you have a specific question or difficulty, please let us know and we'll try to help.

**Comment by crozone** [comment]
So being able to round-trip `DateTime`'s reliably basically comes boils to:

```csharp
public class UtcDateTimeConverter : ValueConverter<DateTime, DateTime>
{
    public UtcDateTimeConverter()
        : base(
            static datetime => datetime.ToUniversalTime(),
            static datetime => DateTime.SpecifyKind(datetime, DateTimeKind.Utc)
            )
    {
    }
}
```

And then manually applying this to every `DateTime` property everywhere.

EF really needs this as a built-in feature, with the ability to specify this behaviour as a global convention. The ability to actually store a zone offset is often unnecessary, rather we just want the `DateTime` round-tripped reliably.

Even if EF picks up the ability to set a global value converter without the need for all that reflection over entity properties, I think UTC DateTime conversion should still be a built-in feature, since it's a *super* common pitfall and all this boilerplate should ultimately be unnecessary.

**Comment by roji** [comment]
> So being able to round-trip DateTime's reliably basically boils down to:

Note that this isn't what "round-tripping" means - your converter makes it so that all DateTimes are returned with Kind=Utc; in other words, the Kind property of the DateTime is lost.

> I think UTC DateTime conversion should still be a built-in feature, since it's a super common pitfall and all this boilerplate should ultimately be unnecessary.

We would in any case not change the default way that DateTime is written and read, as that would be a big breaking change; in addition, Unspecified DateTimes are returned by the database driver itself (SqlClient), and it's not EF's role to override that decision.

Given that we wouldn't change the default, including a type converter such as what you propose above still wouldn't remove the need to actually configure that converter; because of that, and because the actual converter code is very short/trivial, that seems to be quite low value...

**Comment by ajcvickers** [comment]
Agree with everything @roji said, but if you're interested, the [available built-in conversions, are documented](https://learn.microsoft.com/en-us/ef/core/modeling/value-conversions?tabs=data-annotations#built-in-converters).

**Comment by springy76** [comment]
@crozone just a side note on using `ToUniversalTime()` against `DateTimeKind.Unspecified`

```c#
var unspecified = new DateTime(2024, 05, 02, 12, 0, 0);
Console.WriteLine(unspecified.ToLocalTime());
Console.WriteLine(unspecified.ToUniversalTime());
```

`ToLocalTime()` handles DateTimes of unspecified kind as they would be of universal time.
`ToUniversalTime()` handles DateTimes of unspecified kind as they would be of local time.

I'm quite happy with npgsqls decision to just throw on any non-utc value.

**Comment by MattCost** [comment]
>Unspecified DateTimes are returned by the database driver itself (SqlClient), and it's not EF's role to override that decision.

I know this issue is old and closed, but this way of thinking is silly. Isn't the point of a Data ABSTRACTION Layer to be in between a db and code so the code doesn't have to worry about the database limitations?

The database doesn't need to be a 1:1 mapping to the model, the point of the Database layer is to MAP my model to the database. If EF Core is working against AzureSQL it could create 2 columns for a dataTime, `PropertyName_Value` and `PropertyName_Kind` and transparently 

Starting a brand new project with AzureSQL + EF Core. My model has a dateTime. My code sets a DateTimeKind. EF Core generates a table. EF Core saves the data. EF Core fetches the data. Yet EFCore is limited by the backing database? Bah humbug.

**Comment by Ahmadjarad47** [comment]
public static class UtcDateAnnotation
    {
        private const String IsUtcAnnotation = "IsUtc";
        private static readonly ValueConverter<DateTime, DateTime> UtcConverter =
          new ValueConverter<DateTime, DateTime>(v => v, v => DateTime.SpecifyKind(v, DateTimeKind.Utc));

        private static readonly ValueConverter<DateTime?, DateTime?> UtcNullableConverter =
          new ValueConverter<DateTime?, DateTime?>(v => v, v => v == null ? v : DateTime.SpecifyKind(v.Value, DateTimeKind.Utc));

        public static PropertyBuilder<TProperty> IsUtc<TProperty>(this PropertyBuilder<TProperty> builder, Boolean isUtc = true) =>
          builder.HasAnnotation(IsUtcAnnotation, isUtc);

        public static Boolean IsUtc(this IMutableProperty property) =>
          ((Boolean?)property.FindAnnotation(IsUtcAnnotation)?.Value) ?? true;

        /// <summary>
        /// Make sure this is called after configuring all your entities.
        /// </summary>
        public static void ApplyUtcDateTimeConverter(this ModelBuilder builder)
        {
            foreach (var entityType in builder.Model.GetEntityTypes())
            {
                foreach (var property in entityType.GetProperties())
                {
                    if (!property.IsUtc())
                    {
                        continue;
                    }

                    if (property.ClrType == typeof(DateTime))
                    {
                        property.SetValueConverter(UtcConverter);
                    }

                    if (property.ClrType == typeof(DateTime?))
                    {
                        property.SetValueConverter(UtcNullableConverter);
                    }
                }
            }
        }
    }

**Comment by ApacheTech** [comment]
This is the solution I'm using for EFCore in .NET8. It uses mementos for each type to minimise the overhead of reflection for each query.

```cs
/// <summary>
///       Intercepts EF Core materialisation and forces all DateTime properties to DateTimeKind.Utc.
///       Uses cached, compiled delegates for maximum performance.
/// </summary>
public class DateTimeKindUtcInterceptor : IMaterializationInterceptor
{
	private static readonly ConcurrentDictionary<Type, Action<object>[]> _materialisers = [];

	/// <inheritdoc/>
	public object InitializedInstance(MaterializationInterceptionData materialisationData, object entity)
	{
		var entityType = entity.GetType();
		var actions = _materialisers.GetOrAdd(entityType, CreateMaterialisersForType);

		foreach (var action in actions)
		{
			action(entity);
		}

		return entity;
	}

	/// <summary>
	///   Builds and caches an array of compiled delegates that set each DateTime/DateTime? property on an instance to UTC.
	/// </summary>
	private static Action<object>[] CreateMaterialisersForType(Type entityType)
	{
		var props = entityType
			.GetProperties(BindingFlags.Public | BindingFlags.Instance)
			.Where(p => p.CanRead && p.CanWrite)
			.Where(p => p.PropertyType == typeof(DateTime)
					 || p.PropertyType == typeof(DateTime?))
			.ToArray();

		var entityParam = Expression.Parameter(typeof(object), "entity");
		var typedEntity = Expression.Convert(entityParam, entityType);

		var actions = new List<Action<object>>(props.Length);

		foreach (var prop in props)
		{
			var propertyExpr = Expression.Property(typedEntity, prop);
			var lambda = prop.PropertyType == typeof(DateTime)
				? CreateDateTimeInterceptor(entityParam, propertyExpr)
				: CreateNullableDateTimeInterceptor(entityParam, propertyExpr);
			actions.Add(lambda);
		}

		return [.. actions];
	}

	private static Action<object> CreateDateTimeInterceptor(ParameterExpression entityParam, MemberExpression propertyExpr)
	{
		var specifyMethod = typeof(DateTime).GetMethod(nameof(DateTime.SpecifyKind), [typeof(DateTime), typeof(DateTimeKind)])!;
		var specifyCall = Expression.Call(specifyMethod, propertyExpr, Expression.Constant(DateTimeKind.Utc));
		var assign = Expression.Assign(propertyExpr, specifyCall);
		var lambda = Expression.Lambda<Action<object>>(assign, entityParam).Compile();
		return lambda;
	}

	private static Action<object> CreateNullableDateTimeInterceptor(ParameterExpression entityParam, MemberExpression propertyExpr)
	{
		var nullableValue = Expression.Property(propertyExpr, "Value");
		var specifyMethod = typeof(DateTime).GetMethod(nameof(DateTime.SpecifyKind), [typeof(DateTime), typeof(DateTimeKind)])!;
		var specifyCall = Expression.Call(specifyMethod, nullableValue, Expression.Constant(DateTimeKind.Utc));
		var nullableAssign = Expression.Assign(propertyExpr, Expression.Convert(specifyCall, typeof(DateTime?)));
		var hasValueCheck = Expression.Property(propertyExpr, "HasValue");
		var conditional = Expression.IfThen(hasValueCheck, nullableAssign);
		var lambda = Expression.Lambda<Action<object>>(conditional, entityParam).Compile();
		return lambda;
	}
}
```

This interceptor is added as an `IInterceptor` to the DI.

```cs
services.AddSingleton<IInterceptor, DateTimeKindUtcInterceptor>();
```

And we have a generalised way of adding interceptors with

```cs
services.AddDbContextFactory<DbContext>((services, options) =>
{
	options
		.UseSqlServer(...)
		.AddInterceptors(services.GetServices<IInterceptor>());
});
```
